book/src/designs/0002-parallel-verification.md

```diff
@@ -0,0 +1,426 @@
+* **Before Overwinter:**
+  * Add UTXOs, note commitments, and nullifiers from each block to the
+    `ChainContext` as deltas.
+  * Store the full sets in the `DiskState`, and apply deltas to update them.
+* **Overwinter:**
+  * Additional context required to verify Overwinter (v3) transactions, if any.
+* **Sapling:**
+  * Additional context required to verify Sapling (v4) transactions, if any.
+* **Blossom:**
+  * If the context contains pre-calculated fields that depend on
+    `AveragingWindowTimespan`, re-calculate those fields. See
+    [ZIP-208](https://zips.z.cash/zip-0208#effect-on-difficulty-adjustment)
+    for details.
+  * Note: Zebra can avoid this issue by always verifying the `bits`
+    (difficulty) and `time` fields based on recent block headers, without
+    storing any intermediate values in the context.
+* **Heartwood:**
+  * Heartwood changes the meaning of the `history_root_hash` field in the block
+    header. For Sapling and Blossom, it is the final sapling treestate hash.
+    For Heartwood, it is the root hash of a Merkle Mountain Range (MMR) which
+    commits to various features of the chain history. See
+    [ZIP-221](https://zips.z.cash/zip-0221) for details.
+  * Each new block adds a leaf node, then merges subtree roots, if possible.
+    It generates some new "extra nodes" to bag the subtrees, then calculates
+    the root hash.
+  * To efficiently verify this field, Zebra can store the list of MMR previous
+    subtree roots in the `ChainContext` as an `im::Vector`. The number of
+    subtree roots is at most `log(height - activation_height)`.
+  * As an optimisation, we could also store a list of recent "extra nodes"
+    generated during the bagging process.
+  * Store the full list of previous subtree roots in the `DiskState`. Update by
+    replacing the old `DiskState` list with the new `ChainContext` set.
+  * The Heartwood activation block has an all-zeroes `history_root_hash` field,
+    and an empty list of previous subtree roots.
+
+## Main Chain Tip
+[main-chain-tip]: #main-chain-tip
+
+The main chain tip determines the set of valid, historical Zcash transactions.
+The main chain tip is updated whenever a new block is received:
+
+First, the set of chain tips is updated:
+* if the new block uniquely extends an existing chain tip, that existing chain
+  tip is replaced by the new block.
+* if the new block forks from the ancestor of an existing tip, the new block
+  is added to the set of chain tips.
+
+Then, the new main tip is selected, according to these rules:
+* the main tip is the chain tip with the greatest cumulative proof of work,
+  calculated according to the Zcash Specification. (This is a consensus rule.)
+* as a tie-breaker, if multiple chain tips have equal cumulative work, and one
+  of those tips is the current main tip, the main tip does not change. (This is
+  *not* a consensus rule, because it depends on download and verification
+  order on each local node.)
```

Comment:
```suggestion
* as a tie-breaker, if multiple chain tips have equal cumulative work, and one
  of those tips is the current main tip, the main tip does not change. This
  check can be implemented using a strictly greater than comparison. (This
  is *not* a consensus rule, because it depends on download and
  verification order on each local node. But it is an important feature of node
  implementations, because it helps the network converge on a single
  chain.)
```